// @grapesjs/react/dist/index.js
import ze, { createContext as He, useState as x, useContext as Ke, useRef as he, useEffect as C, memo as A, useMemo as Ne, createElement as Ve } from "react";
import { createPortal as Tt } from "react-dom";
var Ee = { exports: {} }, X = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var Ue;
function wt() {
  if (Ue)
    return X;
  Ue = 1;
  var n = ze, t = Symbol.for("react.element"), i = Symbol.for("react.fragment"), o = Object.prototype.hasOwnProperty, s = n.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, c = { key: !0, ref: !0, __self: !0, __source: !0 };
  function u(l, d, y) {
    var h, P = {}, T = null, E = null;
    y !== void 0 && (T = "" + y), d.key !== void 0 && (T = "" + d.key), d.ref !== void 0 && (E = d.ref);
    for (h in d)
      o.call(d, h) && !c.hasOwnProperty(h) && (P[h] = d[h]);
    if (l && l.defaultProps)
      for (h in d = l.defaultProps, d)
        P[h] === void 0 && (P[h] = d[h]);
    return { $$typeof: t, type: l, key: T, ref: E, props: P, _owner: s.current };
  }
  return X.Fragment = i, X.jsx = u, X.jsxs = u, X;
}
var Z = {};
/**
 * @license React
 * react-jsx-runtime.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var qe;
function Ot() {
  return qe || (qe = 1, process.env.NODE_ENV !== "production" && function() {
    var n = ze, t = Symbol.for("react.element"), i = Symbol.for("react.portal"), o = Symbol.for("react.fragment"), s = Symbol.for("react.strict_mode"), c = Symbol.for("react.profiler"), u = Symbol.for("react.provider"), l = Symbol.for("react.context"), d = Symbol.for("react.forward_ref"), y = Symbol.for("react.suspense"), h = Symbol.for("react.suspense_list"), P = Symbol.for("react.memo"), T = Symbol.for("react.lazy"), E = Symbol.for("react.offscreen"), F = Symbol.iterator, N = "@@iterator";
    function J(e) {
      if (e === null || typeof e != "object")
        return null;
      var r = F && e[F] || e[N];
      return typeof r == "function" ? r : null;
    }
    var W = n.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    function _(e) {
      {
        for (var r = arguments.length, a = new Array(r > 1 ? r - 1 : 0), f = 1; f < r; f++)
          a[f - 1] = arguments[f];
        Q("error", e, a);
      }
    }
    function Q(e, r, a) {
      {
        var f = W.ReactDebugCurrentFrame, m = f.getStackAddendum();
        m !== "" && (r += "%s", a = a.concat([m]));
        var g = a.map(function(p) {
          return String(p);
        });
        g.unshift("Warning: " + r), Function.prototype.apply.call(console[e], console, g);
      }
    }
    var ee = !1, G = !1, ue = !1, ce = !1, z = !1, O;
    O = Symbol.for("react.module.reference");
    function te(e) {
      return !!(typeof e == "string" || typeof e == "function" || e === o || e === c || z || e === s || e === y || e === h || ce || e === E || ee || G || ue || typeof e == "object" && e !== null && (e.$$typeof === T || e.$$typeof === P || e.$$typeof === u || e.$$typeof === l || e.$$typeof === d || // This needs to include all possible module reference object
      // types supported by any Flight configuration anywhere since
      // we don't know which Flight build this will end up being used
      // with.
      e.$$typeof === O || e.getModuleId !== void 0));
    }
    function re(e, r, a) {
      var f = e.displayName;
      if (f)
        return f;
      var m = r.displayName || r.name || "";
      return m !== "" ? a + "(" + m + ")" : a;
    }
    function H(e) {
      return e.displayName || "Context";
    }
    function I(e) {
      if (e == null)
        return null;
      if (typeof e.tag == "number" && _("Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."), typeof e == "function")
        return e.displayName || e.name || null;
      if (typeof e == "string")
        return e;
      switch (e) {
        case o:
          return "Fragment";
        case i:
          return "Portal";
        case c:
          return "Profiler";
        case s:
          return "StrictMode";
        case y:
          return "Suspense";
        case h:
          return "SuspenseList";
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case l:
            var r = e;
            return H(r) + ".Consumer";
          case u:
            var a = e;
            return H(a._context) + ".Provider";
          case d:
            return re(e, e.render, "ForwardRef");
          case P:
            var f = e.displayName || null;
            return f !== null ? f : I(e.type) || "Memo";
          case T: {
            var m = e, g = m._payload, p = m._init;
            try {
              return I(p(g));
            } catch {
              return null;
            }
          }
        }
      return null;
    }
    var j = Object.assign, B = 0, Se, be, Re, Ce, _e, Pe, Te;
    function we() {
    }
    we.__reactDisabledLog = !0;
    function tt() {
      {
        if (B === 0) {
          Se = console.log, be = console.info, Re = console.warn, Ce = console.error, _e = console.group, Pe = console.groupCollapsed, Te = console.groupEnd;
          var e = {
            configurable: !0,
            enumerable: !0,
            value: we,
            writable: !0
          };
          Object.defineProperties(console, {
            info: e,
            log: e,
            warn: e,
            error: e,
            group: e,
            groupCollapsed: e,
            groupEnd: e
          });
        }
        B++;
      }
    }
    function rt() {
      {
        if (B--, B === 0) {
          var e = {
            configurable: !0,
            enumerable: !0,
            writable: !0
          };
          Object.defineProperties(console, {
            log: j({}, e, {
              value: Se
            }),
            info: j({}, e, {
              value: be
            }),
            warn: j({}, e, {
              value: Re
            }),
            error: j({}, e, {
              value: Ce
            }),
            group: j({}, e, {
              value: _e
            }),
            groupCollapsed: j({}, e, {
              value: Pe
            }),
            groupEnd: j({}, e, {
              value: Te
            })
          });
        }
        B < 0 && _("disabledDepth fell below zero. This is a bug in React. Please file an issue.");
      }
    }
    var le = W.ReactCurrentDispatcher, fe;
    function ne(e, r, a) {
      {
        if (fe === void 0)
          try {
            throw Error();
          } catch (m) {
            var f = m.stack.trim().match(/\n( *(at )?)/);
            fe = f && f[1] || "";
          }
        return `
` + fe + e;
      }
    }
    var de = !1, oe;
    {
      var nt = typeof WeakMap == "function" ? WeakMap : Map;
      oe = new nt();
    }
    function Oe(e, r) {
      if (!e || de)
        return "";
      {
        var a = oe.get(e);
        if (a !== void 0)
          return a;
      }
      var f;
      de = !0;
      var m = Error.prepareStackTrace;
      Error.prepareStackTrace = void 0;
      var g;
      g = le.current, le.current = null, tt();
      try {
        if (r) {
          var p = function() {
            throw Error();
          };
          if (Object.defineProperty(p.prototype, "props", {
            set: function() {
              throw Error();
            }
          }), typeof Reflect == "object" && Reflect.construct) {
            try {
              Reflect.construct(p, []);
            } catch ($) {
              f = $;
            }
            Reflect.construct(e, [], p);
          } else {
            try {
              p.call();
            } catch ($) {
              f = $;
            }
            e.call(p.prototype);
          }
        } else {
          try {
            throw Error();
          } catch ($) {
            f = $;
          }
          e();
        }
      } catch ($) {
        if ($ && f && typeof $.stack == "string") {
          for (var v = $.stack.split(`
`), w = f.stack.split(`
`), S = v.length - 1, b = w.length - 1; S >= 1 && b >= 0 && v[S] !== w[b]; )
            b--;
          for (; S >= 1 && b >= 0; S--, b--)
            if (v[S] !== w[b]) {
              if (S !== 1 || b !== 1)
                do
                  if (S--, b--, b < 0 || v[S] !== w[b]) {
                    var k = `
` + v[S].replace(" at new ", " at ");
                    return e.displayName && k.includes("<anonymous>") && (k = k.replace("<anonymous>", e.displayName)), typeof e == "function" && oe.set(e, k), k;
                  }
                while (S >= 1 && b >= 0);
              break;
            }
        }
      } finally {
        de = !1, le.current = g, rt(), Error.prepareStackTrace = m;
      }
      var U = e ? e.displayName || e.name : "", Be = U ? ne(U) : "";
      return typeof e == "function" && oe.set(e, Be), Be;
    }
    function ot(e, r, a) {
      return Oe(e, !1);
    }
    function at(e) {
      var r = e.prototype;
      return !!(r && r.isReactComponent);
    }
    function ae(e, r, a) {
      if (e == null)
        return "";
      if (typeof e == "function")
        return Oe(e, at(e));
      if (typeof e == "string")
        return ne(e);
      switch (e) {
        case y:
          return ne("Suspense");
        case h:
          return ne("SuspenseList");
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case d:
            return ot(e.render);
          case P:
            return ae(e.type, r, a);
          case T: {
            var f = e, m = f._payload, g = f._init;
            try {
              return ae(g(m), r, a);
            } catch {
            }
          }
        }
      return "";
    }
    var se = Object.prototype.hasOwnProperty, xe = {}, je = W.ReactDebugCurrentFrame;
    function ie(e) {
      if (e) {
        var r = e._owner, a = ae(e.type, e._source, r ? r.type : null);
        je.setExtraStackFrame(a);
      } else
        je.setExtraStackFrame(null);
    }
    function st(e, r, a, f, m) {
      {
        var g = Function.call.bind(se);
        for (var p in e)
          if (g(e, p)) {
            var v = void 0;
            try {
              if (typeof e[p] != "function") {
                var w = Error((f || "React class") + ": " + a + " type `" + p + "` is invalid; it must be a function, usually from the `prop-types` package, but received `" + typeof e[p] + "`.This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`.");
                throw w.name = "Invariant Violation", w;
              }
              v = e[p](r, p, f, a, null, "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED");
            } catch (S) {
              v = S;
            }
            v && !(v instanceof Error) && (ie(m), _("%s: type specification of %s `%s` is invalid; the type checker function must return `null` or an `Error` but returned a %s. You may have forgotten to pass an argument to the type checker creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and shape all require an argument).", f || "React class", a, p, typeof v), ie(null)), v instanceof Error && !(v.message in xe) && (xe[v.message] = !0, ie(m), _("Failed %s type: %s", a, v.message), ie(null));
          }
      }
    }
    var it = Array.isArray;
    function ve(e) {
      return it(e);
    }
    function ut(e) {
      {
        var r = typeof Symbol == "function" && Symbol.toStringTag, a = r && e[Symbol.toStringTag] || e.constructor.name || "Object";
        return a;
      }
    }
    function ct(e) {
      try {
        return ke(e), !1;
      } catch {
        return !0;
      }
    }
    function ke(e) {
      return "" + e;
    }
    function Ae(e) {
      if (ct(e))
        return _("The provided key is an unsupported type %s. This value must be coerced to a string before before using it here.", ut(e)), ke(e);
    }
    var K = W.ReactCurrentOwner, lt = {
      key: !0,
      ref: !0,
      __self: !0,
      __source: !0
    }, Me, De, pe;
    pe = {};
    function ft(e) {
      if (se.call(e, "ref")) {
        var r = Object.getOwnPropertyDescriptor(e, "ref").get;
        if (r && r.isReactWarning)
          return !1;
      }
      return e.ref !== void 0;
    }
    function dt(e) {
      if (se.call(e, "key")) {
        var r = Object.getOwnPropertyDescriptor(e, "key").get;
        if (r && r.isReactWarning)
          return !1;
      }
      return e.key !== void 0;
    }
    function vt(e, r) {
      if (typeof e.ref == "string" && K.current && r && K.current.stateNode !== r) {
        var a = I(K.current.type);
        pe[a] || (_('Component "%s" contains the string ref "%s". Support for string refs will be removed in a future major release. This case cannot be automatically converted to an arrow function. We ask you to manually fix this case by using useRef() or createRef() instead. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-string-ref', I(K.current.type), e.ref), pe[a] = !0);
      }
    }
    function pt(e, r) {
      {
        var a = function() {
          Me || (Me = !0, _("%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", r));
        };
        a.isReactWarning = !0, Object.defineProperty(e, "key", {
          get: a,
          configurable: !0
        });
      }
    }
    function mt(e, r) {
      {
        var a = function() {
          De || (De = !0, _("%s: `ref` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", r));
        };
        a.isReactWarning = !0, Object.defineProperty(e, "ref", {
          get: a,
          configurable: !0
        });
      }
    }
    var gt = function(e, r, a, f, m, g, p) {
      var v = {
        // This tag allows us to uniquely identify this as a React Element
        $$typeof: t,
        // Built-in properties that belong on the element
        type: e,
        key: r,
        ref: a,
        props: p,
        // Record the component responsible for creating this element.
        _owner: g
      };
      return v._store = {}, Object.defineProperty(v._store, "validated", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: !1
      }), Object.defineProperty(v, "_self", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: f
      }), Object.defineProperty(v, "_source", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: m
      }), Object.freeze && (Object.freeze(v.props), Object.freeze(v)), v;
    };
    function yt(e, r, a, f, m) {
      {
        var g, p = {}, v = null, w = null;
        a !== void 0 && (Ae(a), v = "" + a), dt(r) && (Ae(r.key), v = "" + r.key), ft(r) && (w = r.ref, vt(r, m));
        for (g in r)
          se.call(r, g) && !lt.hasOwnProperty(g) && (p[g] = r[g]);
        if (e && e.defaultProps) {
          var S = e.defaultProps;
          for (g in S)
            p[g] === void 0 && (p[g] = S[g]);
        }
        if (v || w) {
          var b = typeof e == "function" ? e.displayName || e.name || "Unknown" : e;
          v && pt(p, b), w && mt(p, b);
        }
        return gt(e, v, w, m, f, K.current, p);
      }
    }
    var me = W.ReactCurrentOwner, Fe = W.ReactDebugCurrentFrame;
    function V(e) {
      if (e) {
        var r = e._owner, a = ae(e.type, e._source, r ? r.type : null);
        Fe.setExtraStackFrame(a);
      } else
        Fe.setExtraStackFrame(null);
    }
    var ge;
    ge = !1;
    function ye(e) {
      return typeof e == "object" && e !== null && e.$$typeof === t;
    }
    function Ie() {
      {
        if (me.current) {
          var e = I(me.current.type);
          if (e)
            return `

Check the render method of \`` + e + "`.";
        }
        return "";
      }
    }
    function Et(e) {
      {
        if (e !== void 0) {
          var r = e.fileName.replace(/^.*[\\\/]/, ""), a = e.lineNumber;
          return `

Check your code at ` + r + ":" + a + ".";
        }
        return "";
      }
    }
    var Le = {};
    function ht(e) {
      {
        var r = Ie();
        if (!r) {
          var a = typeof e == "string" ? e : e.displayName || e.name;
          a && (r = `

Check the top-level render call using <` + a + ">.");
        }
        return r;
      }
    }
    function We(e, r) {
      {
        if (!e._store || e._store.validated || e.key != null)
          return;
        e._store.validated = !0;
        var a = ht(r);
        if (Le[a])
          return;
        Le[a] = !0;
        var f = "";
        e && e._owner && e._owner !== me.current && (f = " It was passed a child from " + I(e._owner.type) + "."), V(e), _('Each child in a list should have a unique "key" prop.%s%s See https://reactjs.org/link/warning-keys for more information.', a, f), V(null);
      }
    }
    function $e(e, r) {
      {
        if (typeof e != "object")
          return;
        if (ve(e))
          for (var a = 0; a < e.length; a++) {
            var f = e[a];
            ye(f) && We(f, r);
          }
        else if (ye(e))
          e._store && (e._store.validated = !0);
        else if (e) {
          var m = J(e);
          if (typeof m == "function" && m !== e.entries)
            for (var g = m.call(e), p; !(p = g.next()).done; )
              ye(p.value) && We(p.value, r);
        }
      }
    }
    function St(e) {
      {
        var r = e.type;
        if (r == null || typeof r == "string")
          return;
        var a;
        if (typeof r == "function")
          a = r.propTypes;
        else if (typeof r == "object" && (r.$$typeof === d || // Note: Memo only checks outer props here.
        // Inner props are checked in the reconciler.
        r.$$typeof === P))
          a = r.propTypes;
        else
          return;
        if (a) {
          var f = I(r);
          st(a, e.props, "prop", f, e);
        } else if (r.PropTypes !== void 0 && !ge) {
          ge = !0;
          var m = I(r);
          _("Component %s declared `PropTypes` instead of `propTypes`. Did you misspell the property assignment?", m || "Unknown");
        }
        typeof r.getDefaultProps == "function" && !r.getDefaultProps.isReactClassApproved && _("getDefaultProps is only used on classic React.createClass definitions. Use a static property named `defaultProps` instead.");
      }
    }
    function bt(e) {
      {
        for (var r = Object.keys(e.props), a = 0; a < r.length; a++) {
          var f = r[a];
          if (f !== "children" && f !== "key") {
            V(e), _("Invalid prop `%s` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.", f), V(null);
            break;
          }
        }
        e.ref !== null && (V(e), _("Invalid attribute `ref` supplied to `React.Fragment`."), V(null));
      }
    }
    function Ye(e, r, a, f, m, g) {
      {
        var p = te(e);
        if (!p) {
          var v = "";
          (e === void 0 || typeof e == "object" && e !== null && Object.keys(e).length === 0) && (v += " You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.");
          var w = Et(m);
          w ? v += w : v += Ie();
          var S;
          e === null ? S = "null" : ve(e) ? S = "array" : e !== void 0 && e.$$typeof === t ? (S = "<" + (I(e.type) || "Unknown") + " />", v = " Did you accidentally export a JSX literal instead of a component?") : S = typeof e, _("React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: %s.%s", S, v);
        }
        var b = yt(e, r, a, m, g);
        if (b == null)
          return b;
        if (p) {
          var k = r.children;
          if (k !== void 0)
            if (f)
              if (ve(k)) {
                for (var U = 0; U < k.length; U++)
                  $e(k[U], e);
                Object.freeze && Object.freeze(k);
              } else
                _("React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead.");
            else
              $e(k, e);
        }
        return e === o ? bt(b) : St(b), b;
      }
    }
    function Rt(e, r, a) {
      return Ye(e, r, a, !0);
    }
    function Ct(e, r, a) {
      return Ye(e, r, a, !1);
    }
    var _t = Ct, Pt = Rt;
    Z.Fragment = o, Z.jsx = _t, Z.jsxs = Pt;
  }()), Z;
}
process.env.NODE_ENV === "production" ? Ee.exports = wt() : Ee.exports = Ot();
var R = Ee.exports;
const Xe = He(null), xt = ({ children: n }) => {
  const [t, i] = x({
    setEditor: (o) => {
      i((s) => ({ ...s, editor: o }));
    }
  });
  return /* @__PURE__ */ R.jsx(Xe.Provider, { value: t, children: n });
}, M = () => {
  const n = Ke(Xe);
  if (!n)
    throw new Error("useEditorInstance must be used within EditorInstanceProvider");
  return n;
}, Nt = () => {
  const { editor: n } = M();
  if (!n)
    throw new Error("useEditor used before the load of the editor instance. You can wrap your component in `<WithEditor>` or make use of `useEditorMaybe` hook and ensure the `editor` exists.");
  return n;
}, jt = () => M().editor, Ze = He(null), kt = ({ children: n }) => {
  const [t, i] = x(() => ({
    setRefCanvas(o) {
      i((s) => ({ ...s, refCanvas: o }));
    },
    setCustomModal(o) {
      i((s) => ({ ...s, customModal: o }));
    },
    setCustomAssets(o) {
      i((s) => ({ ...s, customAssets: o }));
    },
    setCustomBlocks(o) {
      i((s) => ({ ...s, customBlocks: o }));
    },
    setCustomRte(o) {
      i((s) => ({ ...s, customRte: o }));
    },
    setCustomStyles(o) {
      i((s) => ({ ...s, customStyles: o }));
    },
    setCustomLayers(o) {
      i((s) => ({ ...s, customLayers: o }));
    },
    setCustomSelectors(o) {
      i((s) => ({ ...s, customSelectors: o }));
    },
    setCustomTraits(o) {
      i((s) => ({ ...s, customTraits: o }));
    },
    setReady(o) {
      i((s) => ({ ...s, ready: o }));
    }
  }));
  return /* @__PURE__ */ R.jsx(Ze.Provider, { value: t, children: n });
}, L = () => {
  const n = Ke(Ze);
  if (!n)
    throw new Error("useEditorOptions must be used within EditorOptionsProvider");
  return n;
};
function At(...n) {
  return (Array.isArray(n[0]) ? n[0] : [...n]).filter(Boolean).join(" ");
}
function Y(n) {
  return typeof n == "function";
}
function D() {
}
const Qe = (n) => typeof n == "string", Mt = async (n) => {
  const t = document.createElement("link");
  t.href = n, t.rel = "stylesheet", document.head.appendChild(t);
}, et = (n) => {
  const t = Qe(n) ? { id: n, src: n } : n;
  return new Promise((i, o) => {
    const s = document.createElement("script");
    if (document.querySelector(`script[src="${t.src}"]`))
      return i(t.id);
    s.src = t.src, s.onload = () => i(t.id), s.onerror = () => o(t.id), document.head.appendChild(s);
  });
}, Dt = (n) => {
  const t = n.map((i) => et(i));
  return Promise.allSettled(t);
}, Je = (n) => function() {
  const i = he(null);
  return C(() => {
    const { current: o } = i;
    o && (Qe(n) ? o.innerHTML = n : o.appendChild(n));
  }, [i.current]), /* @__PURE__ */ R.jsx("div", { ref: i });
}, Ge = /* @__PURE__ */ new WeakMap();
function q(n) {
  if (!n)
    return () => null;
  const t = Ge.get(n);
  if (t)
    return t;
  const i = function({ children: s }) {
    return Tt(s, n);
  };
  return Ge.set(n, i), i;
}
const Vt = A(function({ children: n }) {
  const { editor: t } = M(), i = L(), [o, s] = x(!1), [c, u] = x({
    assets: [],
    types: [],
    close: () => {
    },
    select: () => {
    },
    Container: () => null
  });
  return C(() => {
    if (!t)
      return;
    const l = t.Assets.events.custom, d = ({ open: y, assets: h, types: P, select: T, close: E, container: F }) => {
      y && u({
        assets: h,
        types: P,
        select: T,
        close: E,
        Container: q(F)
      }), s(y);
    };
    return t.on(l, d), () => {
      t.off(l, d);
    };
  }, [t]), C(() => i.setCustomAssets(!0), []), t && Y(n) ? n({ open: o, ...c }) : null;
}), Ut = A(function({ children: n }) {
  const { editor: t } = M(), i = L(), [o, s] = x(() => ({
    blocks: [],
    dragStart: D,
    dragStop: D,
    mapCategoryBlocks: /* @__PURE__ */ new Map(),
    Container: () => null
  }));
  return C(() => {
    if (!t)
      return;
    const c = t.Blocks.events.custom, u = ({ blocks: l, container: d, dragStart: y, dragStop: h }) => {
      const P = l.reduce((T, E) => {
        const F = E.getCategoryLabel(), N = T.get(F);
        return N ? N.push(E) : T.set(F, [E]), T;
      }, /* @__PURE__ */ new Map());
      s({
        blocks: l,
        dragStart: y,
        dragStop: h,
        mapCategoryBlocks: P,
        Container: q(d)
      });
    };
    return t.on(c, u), t.Blocks.__trgCustom(), () => {
      t.off(c, u);
    };
  }, [t]), C(() => i.setCustomBlocks(!0), []), t && Y(n) ? n(o) : null;
});
function qt({ children: n, ...t }) {
  const i = L(), o = he(null);
  return C(() => {
    o.current && i.setRefCanvas(o.current);
  }, [o.current]), /* @__PURE__ */ R.jsx("div", { ...t, ref: o, children: n });
}
const Jt = A(function({ children: n }) {
  const { editor: t } = M(), [i, o] = x(() => ({
    devices: [],
    selected: "",
    select: D
  }));
  return C(() => {
    if (!t)
      return;
    const { Devices: s } = t, c = s.events.all, u = () => {
      var l;
      o({
        devices: s.getDevices(),
        selected: (l = s.getSelected()) == null ? void 0 : l.id,
        select: (d) => s.select(d)
      });
    };
    return t.on(c, u), u(), () => {
      t.off(c, u);
    };
  }, [t]), t && Y(n) ? n(i) : null;
}), Ft = (n) => !!(n && !Array.isArray(n) && typeof n == "object");
async function It(n) {
  const t = n.map(({ id: u, src: l }) => ({ id: u, src: l })), i = n.reduce((u, l) => (u[l.id] = l, u), {}), o = [], s = [];
  return (await Dt(t)).forEach((u) => {
    u.status === "fulfilled" ? o.push(i[u.value]) : s.push(i[u.reason]);
  }), { loaded: o, failed: s };
}
const Lt = async (n) => {
  const t = [...n], i = {};
  if (t.length) {
    const o = {}, s = [];
    if (t.forEach((c, u) => {
      Ft(c) && (o[c.id] = { index: u }, s.push(c));
    }), s.length) {
      const { loaded: c } = await It(s);
      c.forEach(({ id: u, options: l }) => {
        o[u].loaded = !0, i[u] = l || {};
      });
    }
    Object.keys(o).forEach((c) => {
      const u = o[c];
      u.loaded ? t[u.index] = c : t[u.index] = !1;
    });
  }
  return {
    plugins: t.filter(Boolean),
    pluginOptions: i
  };
}, Wt = A(function({
  children: n,
  className: t,
  style: i,
  options: o = {},
  plugins: s = [],
  grapesjs: c,
  grapesjsCss: u,
  onEditor: l = D,
  onReady: d,
  onUpdate: y,
  waitReady: h,
  ...P
}) {
  const { setEditor: T } = M(), E = L(), [F, N] = x(!1), J = he(null);
  C(() => {
    if (!E.ready || !J.current)
      return;
    const ce = J.current, z = E.refCanvas;
    let O, te = {}, re = [];
    const H = (j) => {
      const B = {
        height: "100%",
        ...o,
        plugins: [
          ...re,
          ...o.plugins || []
        ],
        pluginsOpts: {
          ...o.pluginsOpts,
          ...te
        },
        modal: {
          ...o.modal,
          custom: !!E.customModal
        },
        assetManager: {
          ...o.assetManager,
          custom: !!E.customAssets
        },
        styleManager: {
          ...o.styleManager,
          custom: !!E.customStyles
        },
        blockManager: {
          ...o.blockManager,
          custom: !!E.customBlocks
        },
        richTextEditor: {
          ...o.richTextEditor,
          custom: !!E.customRte
        },
        layerManager: {
          ...o.layerManager,
          custom: !!E.customLayers
        },
        traitManager: {
          ...o.traitManager,
          custom: !!E.customTraits
        },
        selectorManager: {
          ...o.selectorManager,
          custom: !!E.customSelectors
        },
        container: z || ce,
        customUI: !!z,
        // Disables all default panels if Canvas is used
        ...z ? {
          panels: { defaults: [] }
        } : {}
      };
      O = j.init(B), T(O), l(O), y && O.on("update", () => {
        y(O.getProjectData(), O);
      }), O.onReady(() => {
        N(!0), d == null || d(O);
      });
    };
    return (async () => {
      u && await Mt(u);
      const j = await Lt(s);
      re = j.plugins, te = j.pluginOptions, typeof c == "string" ? (await et(c), H(window.grapesjs)) : H(c);
    })(), () => O == null ? void 0 : O.destroy();
  }, [E.ready]);
  const W = o.height ?? "100%", _ = o.width ?? "100%", Q = At("gjs-editor-wrapper", t), ee = h && !F, G = Ne(() => ({
    ...i,
    height: W,
    width: _
  }), [W, _, i]), ue = Ne(() => ({
    ...G,
    ...ee ? {
      opacity: 0,
      width: 0,
      height: 0
    } : {}
  }), [G, ee]);
  return /* @__PURE__ */ R.jsxs(R.Fragment, { children: [
    h && !F ? /* @__PURE__ */ R.jsx("div", { className: Q, style: G, children: h }) : null,
    /* @__PURE__ */ R.jsx("div", { ...P, ref: J, className: Q, style: ue, children: n })
  ] });
}), $t = A(function() {
  const n = L();
  return C(() => n.setReady(!0), []), /* @__PURE__ */ R.jsx(R.Fragment, {});
}), Gt = A(function({ children: t, ...i }) {
  return /* @__PURE__ */ R.jsx(xt, { children: /* @__PURE__ */ R.jsx(kt, { children: /* @__PURE__ */ R.jsxs(Wt, { ...i, children: [
    t,
    /* @__PURE__ */ R.jsx($t, {})
  ] }) }) });
}), zt = A(function({ children: n }) {
  const { editor: t } = M(), i = L(), [o, s] = x(() => ({
    root: void 0,
    Container: () => null
  }));
  return C(() => {
    if (!t)
      return;
    const { Layers: c } = t, u = c.events.custom, l = ({ container: d }) => {
      s({
        root: c.getRoot(),
        Container: q(d)
      });
    };
    return t.on(u, l), c.__trgCustom({}), () => {
      t.off(u, l);
    };
  }, [t]), C(() => i.setCustomLayers(!0), []), t && Y(n) ? n(o) : null;
}), Ht = A(function({ children: n }) {
  const { editor: t } = M(), i = L(), [o, s] = x(!1), [c, u] = x({
    title: /* @__PURE__ */ R.jsx(R.Fragment, {}),
    content: /* @__PURE__ */ R.jsx(R.Fragment, {}),
    attributes: {},
    close: D
  });
  return C(() => {
    if (!t)
      return;
    const l = "modal", d = ({ open: y, title: h, content: P, close: T, attributes: E }) => {
      y && u({
        title: Ve(Je(h)),
        content: Ve(Je(P)),
        attributes: E,
        close: T
      }), s(y);
    };
    return t.on(l, d), () => {
      t.off(l, d);
    };
  }, [t]), C(() => i.setCustomModal(!0), []), t && typeof n == "function" ? n({ open: o, ...c }) : null;
}), Kt = A(function({ children: n }) {
  const { editor: t } = M(), [i, o] = x(() => ({
    pages: [],
    selected: void 0,
    select: D,
    add: D,
    remove: D
  }));
  return C(() => {
    if (!t)
      return;
    const { Pages: s } = t, c = s.events.all, u = () => {
      o({
        pages: s.getAll(),
        selected: s.getSelected(),
        select: (...l) => s.select(...l),
        add: (...l) => s.add(...l),
        remove: (...l) => s.remove(...l)
      });
    };
    return t.on(c, u), u(), () => {
      t.off(c, u);
    };
  }, [t]), t && Y(n) ? n(i) : null;
}), Xt = A(function({ children: n }) {
  const { editor: t } = M(), i = L(), [o, s] = x(() => ({
    selectors: [],
    states: [],
    selectedState: "",
    targets: [],
    addSelector: D,
    removeSelector: D,
    setState: D,
    Container: () => null
  }));
  return C(() => {
    if (!t)
      return;
    const { Selectors: c } = t, u = c.events.custom, l = ({ container: d }) => {
      s({
        selectors: c.getSelected(),
        states: c.getStates(),
        selectedState: c.getState(),
        targets: c.getSelectedTargets().map((y) => y.getSelectorsString()),
        addSelector: (...y) => c.addSelected(...y),
        removeSelector: (...y) => c.removeSelected(...y),
        setState: (...y) => c.setState(...y),
        Container: q(d)
      });
    };
    return t.on(u, l), () => {
      t.off(u, l);
    };
  }, [t]), C(() => i.setCustomSelectors(!0), []), t && Y(n) ? n(o) : null;
}), Zt = A(function({ children: n }) {
  const { editor: t } = M(), i = L(), [o, s] = x(() => ({
    sectors: [],
    Container: () => null
  }));
  return C(() => {
    if (!t)
      return;
    const { Styles: c } = t, u = c.events.custom, l = ({ container: d }) => {
      s({
        sectors: c.getSectors({ visible: !0 }),
        Container: q(d)
      });
    };
    return t.on(u, l), () => {
      t.off(u, l);
    };
  }, [t]), C(() => i.setCustomStyles(!0), []), t && Y(n) ? n(o) : null;
}), Qt = A(function({ children: n }) {
  const { editor: t } = M(), i = L(), [o, s] = x(() => ({
    traits: [],
    Container: () => null
  }));
  return C(() => {
    if (!t)
      return;
    const { Traits: c } = t, u = c.events.custom, l = ({ container: d }) => {
      s({
        traits: c.getCurrent(),
        Container: q(d)
      });
    };
    return t.on(u, l), c.__trgCustom(), () => {
      t.off(u, l);
    };
  }, [t]), C(() => i.setCustomTraits(!0), []), t && Y(n) ? n(o) : null;
}), er = ({ children: n }) => jt() ? /* @__PURE__ */ R.jsx(R.Fragment, { children: n }) : null;
export {
  Vt as AssetsProvider,
  Ut as BlocksProvider,
  qt as Canvas,
  Jt as DevicesProvider,
  Gt as Editor,
  zt as LayersProvider,
  Ht as ModalProvider,
  Kt as PagesProvider,
  Xt as SelectorsProvider,
  Zt as StylesProvider,
  Qt as TraitsProvider,
  er as WithEditor,
  Gt as default,
  Nt as useEditor,
  jt as useEditorMaybe
};
